c++/src/kj/string.h

```diff
@@ -504,7 +648,44 @@ String strArray(T&& arr, const char* delim) {
   return result;
 }
 
-template <typename... Params>
+enum class StringifyContext {
+  NORMAL,
+  // strPreallocated is being called outside of a signal handler (i.e. just for optimizations
+  // purposes to avoid allocations for the scratch buffer) and so allocations aren't fatal.
+  SIGNAL_HANDLER,
+  // strPreallocated is being called within a signal handler. Refuse to stringify types that aren't
+  // on an explicit allow list.
+};
+
+template <typename T>
+static constexpr inline bool isSignalSafeToCharSequence() {
+  return _::IsSignalSafeToCharSequence_<T>::value;
+}
+// Returns true if T is safe to stringify in a signal handler context. This means no memory
+// allocations and no invocations of methods that are not safe not use in a signal handler context.
+// These are manually annotated for each type and this must be done carefully to avoid accidentally
+// allowing types to be passed to strPreallocated.
+
+namespace _ {   // private
+template <StringifyContext Context>
+static inline constexpr void assertSignalSafeToCharSequence() {}
+template <StringifyContext Context, typename T>
+static inline constexpr void assertSignalSafeToCharSequence() {
+  // This is broken out into a separate function so that the compile error will tell you exactly
+  // the type that was problematic.
+  static_assert((Context == StringifyContext::NORMAL) || isSignalSafeToCharSequence<T>(),
+      "Not safe to stringify in a signal handler.");
+}
+template <StringifyContext Context, typename T, typename U, typename ... Rest>
```

Comment:
You should only need two specializations for induction here: a zero-arg specialization and a one-or-more-arg specialization. You seem to have three: a zero-arg, a 1-arg, and a 2-or-more-arg...

Reply:
The purpose of the single argument variant is to ensure that the compiler tells you the exact type that it's having a problem with. If you have the one-or-more arg specialization, then you lose that... but maybe it's OK because it'll be the first argument in the "rest"...